### DISCUSSION TITLE
Outsourcing AlpineJS Modules

### DISCUSSION THREAD

**Discussion by simonferndriger** [alpinejs/alpine#3529] [1. Help] [answered]
Just wondering if it is possible to outsource certain parts of the html that is not always used into a module and only load when needed (for example an about dialog)?

I tried `<div x-data="{ test: '<div x-data=´{ newtest: 'awesome' }´ x-text=´newtest´></div>' }" x-html="test">doesn't work</div>`, however, it doesn't look like Alpine code is interpreted correctly when injected with `x-html` like this, because I got the error that `awesome` is an unexpected identifier. Is there another way?

**Comment by SimoTod** [comment]
X-if

**Reply by simonferndriger** [reply]
You mean `<template x-data="{ fetchedData: fetch(...) }" x-if="fetchedData" x-html="fetchedData">...</template>`?

**Reply by SimoTod** [reply, marked as answer]
Not really. No x-data on a template because that's not really part of the rendered page.
x-if is to defer a portion of HTML until a certain event occurs (i.e, dialog opened). The way you load the inner content is up to you but a lot of people use x-init. For example: https://codepen.io/SimoTod/pen/eYPZrwR

**Reply by ekwoka** [reply]
Similar to the above, there is a plugin that can help with this: https://www.npmjs.com/package/@ekwoka/x-ajax if that's a route you want to go.

**Reply by simonferndriger** [reply]
Just explored that you can't load a `<script src="...">` within the html that is injected via *x-ajax*. Is this behaviour on purpose? My use case is loading a HubSpot form via JS snippet.

**Reply by SimoTod** [reply]
It's in the html specs. Script tags injected dynamically in that way do not run: https://www.w3.org/TR/2008/WD-html5-20080610/dom.html#innerhtml0

**Reply by ekwoka** [reply]
You need to loop over then make new script elements, assign the attributes individually and then append that.


Or just directly import the src

**Reply by simonferndriger** [reply]
Thanks!

**Comment by ekwoka** [comment]
Alpine doesn't have an SPA mode.

Trying to make it be an SPA is likely to be a headache.

If you want full JS templating like that, use something made for it, like Solidjs.

**Reply by simonferndriger** [reply]
Thank you for the hint. However, it is only to be going a FewPageApp, so I guess it will work.
